core/base/scalarFieldCriticalPoints/ScalarFieldCriticalPoints.h

```diff
@@ -483,6 +494,42 @@ int ttk::ScalarFieldCriticalPoints::getNumberOfLowerUpperComponents(
     upperList[i] = upperList[i]->find();
 
   std::vector<UnionFind *>::iterator it;
+  // We retrieve the lower and upper components if we want them
+  if(upperComponents != nullptr && lowerComponents != nullptr) {
+    std::vector<UnionFind *> upperUnionFind;
+    upperComponents->push_back(
+      std::vector<ttk::SimplexId>(1, upperNeighbors[0]));
+    upperUnionFind.push_back(upperList[0]);
+    for(ttk::SimplexId i = 1; i < (SimplexId)upperNeighbors.size(); i++) {
+      it
+        = std::find(upperUnionFind.begin(), upperUnionFind.end(), upperList[i]);
```

Comment:
I believe this piece of code can be much simpler.
after line 494, you just need to iterate on `upperNeighbors` (each entry is a local vertex identifier). For the entry i (i.e. for the vertex `upperNeighbors[i]`), its corresponding union-find seed is given by `upperList[i]->find()`. at this point you may want to store the current vertex (`upperNeighbors[i]`) in a map of vectors (`std::unordered_map<std::vector<SimplexId>, SimplexId>`) which stores for each component id, its list of vertices.

Reply:
the suggested usage of a map will trade the quadratic loop (find after insertions) by something faster in practice.